core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/search/service/solr/indexer/AbstractSolrIndexUpdateServiceImpl.java

```diff
@@ -0,0 +1,292 @@
+                                stopRunning();
+                                return;
+                            }
+                            
+                            command = (SolrUpdateCommand)queue.poll(getQueuePollTime(), TimeUnit.MILLISECONDS);
+                            
+                            if (command != null) {
+                                try {
+                                    //We're running in a background thread, so let's just set up a new BroadleafRequestContext.
+                                    BroadleafRequestContext.setBroadleafRequestContext(new BroadleafRequestContext());
+                                    commandHandler.executeCommand(command);
+                                } catch (Exception e) {
+                                    LOG.error("Unexpected error occured attempting to update a Solr index.", e);
+                                } finally {
+                                    BroadleafRequestContext.setBroadleafRequestContext(null);
+                                }
+                            }
+                        }
+                    } finally {
+                        lock.unlock();
+                    }
+                    
+                }
+                
+            } catch (InterruptedException e) {
+                stopRunning();
+                return;
+            }
+            
+        }
+        
+        public synchronized boolean isRunning() {
+            return running;
+        }
+        
+        public synchronized void stopRunning() {
+            this.running = false;
+        }
+        
+        private synchronized void startRunning() {
+            this.running = true;
+        }
+    }
+    
+    public static boolean isRunning(String commandIdentifier) {
+        synchronized (AbstractSolrIndexUpdateServiceImpl.class) {
+            return commandThreadRegistry.containsKey(commandIdentifier) && ((CommandCoordinator)commandThreadRegistry.get(commandIdentifier).get(0)).isRunning();
+            
+        }
+    }
+    
+    @Override
+    public void rebuildIndex() throws ServiceException {
+        scheduleCommand(FullReindexCommand.DEFAULT_INSTANCE);
+    }
+    
+    @Override
+    public void updateIndex(List<SolrInputDocument> documents) {
+        updateIndex(documents, null);
+    }
+
+    @Override
+    public void updateIndex(List<SolrInputDocument> documents, List<String> deleteQueries) {
+        IncrementalUpdateCommand cmd = new IncrementalUpdateCommand(documents, deleteQueries);
+        scheduleCommand(cmd);
+    }
+    
+    @Override
+    public SolrInputDocument buildDocument(Indexable indexable) {
+        return commandHandler.buildDocument(indexable);
+    }
+    
+    @Override
+    public SolrInputDocument buildDocument(Indexable indexable, List<IndexField> fields, List<Locale> locales) {
+        return commandHandler.buildDocument(indexable, fields, locales);
+    }
```

Comment:
It seems like these `buildDocument(...)` methods don't belong in this class - I'd expect to only see them in the CommandHandler. @ktisdell do you have a particular use case in mind for these methods?

Reply:
@ckittrell they don't need to be.  I originally added them so that services outside of this class can build documents as needed.  Consider, for example, `WorkflowSolrIndexServiceImpl#getSolrDocuments(...)`.  In fact, this should likely be updated to optionally use this class for building documents rather than using the legacy class.

Reply:
Although, they could likely just use the `CatalogDocumentBuilder` as well.